xarray/core/combine.py

```diff
@@ -421,6 +45,71 @@ def _infer_tile_ids_from_nested_list(entry, current_pos):
         yield current_pos, entry
 
 
+def _infer_concat_order_from_coords(datasets):
+
+    concat_dims = []
+    tile_ids = [() for ds in datasets]
+
+    # All datasets have same variables because they've been grouped as such
+    ds0 = datasets[0]
+    for dim in ds0.dims:
+
+        # Check if dim is a coordinate dimension
+        if dim in ds0:
+
+            # Need to read coordinate values to do ordering
+            indexes = [ds.indexes.get(dim) for ds in datasets]
+            if any(index is None for index in indexes):
+                raise ValueError("Every dimension needs a coordinate for "
+                                 "inferring concatenation order")
+
+            # If dimension coordinate values are same on every dataset then
+            # should be leaving this dimension alone (it's just a "bystander")
+            if not all(index.equals(indexes[0]) for index in indexes[1:]):
+
+                # Infer order datasets should be arranged in along this dim
+                concat_dims.append(dim)
+
+                if all(index.is_monotonic_increasing for index in indexes):
+                    ascending = True
+                elif all(index.is_monotonic_decreasing for index in indexes):
+                    ascending = False
+                else:
+                    raise ValueError("Coordinate variable {} is neither "
+                                     "monotonically increasing nor "
+                                     "monotonically decreasing on all datasets"
+                                     .format(dim))
+
+                # Assume that any two datasets whose coord along dim starts
+                # with the same value have the same coord values throughout.
+                try:
+                    first_items = pd.Index([index.take([0])
+                                            for index in indexes])
+                except IndexError:
+                    raise ValueError('Cannot handle size zero dimensions')
```

Comment:
This is a minor point, but it would be a little less error prone to explicitly check the size of each sub-index, e.g., `if any(index.size == 0 for index in indexes)`.